packages/filesystem/src/node/disk-file-system-provider.ts

```diff
@@ -509,7 +509,12 @@ export class DiskFileSystemProvider implements Disposable,
             if (opts.recursive) {
                 await this.rimraf(filePath);
             } else {
-                await promisify(unlink)(filePath);
+                const stat = await promisify(lstat)(filePath);
+                if (stat.isDirectory() && !stat.isSymbolicLink()) {
```

Comment:
I am testing on Ubuntu 22.04 and it looks very good. I noticed one weird thing however. 

Given that `/tmp/foo/Untitled` is a directory and a symbolic link created with `ln -s`. 

`lsstat` has different results depending on whether the path ends with a / or without. 
 
with a trailing `/` `stat.isDirectory()` is `true` while `stat.isSymbolicLink()` is `false`. 
without a trailing `/` `stat.isDirectory()` is `false` while `stat.isSymbolicLink()` is `true`. 

This means that
```
                const folderUri = URI.fromFilePath('/tmp/foo/Untitled/');
                this.fileService.delete(folderUri, { recursive: false });
```
fails, while
```
                const folderUri = URI.fromFilePath('/tmp/foo/Untitled');
                this.fileService.delete(folderUri, { recursive: false });
```
is working. 

Maybe we have to normalize the file path by removing trailing `/` first?

Do you see this behavior on your machine as well?

Reply:
Yes, I see the same behavior, but I actually also have Ubuntu 22.04 :-).

if symlink.to.test.folder is a symlink to a folder, I would think that 'rm symlink.to.test.folder/'
asks to delete the folder the symlink links to, whereas 'rm symlink.to.test.folder' asks to remove the symbolic link itself. But for example when experimenting with bash:

`d.friederich@xdfriederich:.../packages/filesystem$rmdir /tmp/tt/symlink.to.test.folder/
rmdir: failed to remove '/tmp/tt/symlink.to.test.folder/': Not a directory
d.friederich@xdfriederich:.../packages/filesystem$rm /tmp/tt/symlink.to.test.folder/
rm: cannot remove '/tmp/tt/symlink.to.test.folder/': Is a directory
d.friederich@xdfriederich:.../packages/filesystem$stat /tmp/tt/symlink.to.test.folder/
  File: /tmp/tt/symlink.to.test.folder/
  Size: 4096      	Blocks: 8          IO Block: 4096   directory
Device: 10302h/66306d	Inode: 25834198    Links: 2
Access: (0755/drwxr-xr-x)  Uid: (503046/d.friederich)   Gid: (103210/d.friederich_g)
Access: 2024-02-08 13:42:01.011807685 -0600
Modify: 2024-02-08 13:42:01.011807685 -0600
Change: 2024-02-08 13:42:01.011807685 -0600
 Birth: 2024-02-08 13:42:01.011807685 -0600
`

so for stat and rm, symlink.to.test.folder/ is referencing a directory, but for rmdir it is not a directory.
So a bit inconsistent. I don't think we should drop the trailing / in this particular case, IMHO asking to delete symlink.to.test.folder/ is not the same as asking to delete symlink.to.test.folder
The current behavior is following what the command line tools do, it fails. In this particular case, both rm and rmdir are failing.

 (The ubuntu rm actually has an option to make it allow to remove an empty directory -d, but in this case it fails because it is not a directory. so rm is inconsistent within itself :-)
 `d.friederich@xdfriederich:.../packages/filesystem$rm /tmp/tt/symlink.to.test.folder/
rm: cannot remove '/tmp/tt/symlink.to.test.folder/': Is a directory
d.friederich@xdfriederich:.../packages/filesystem$rm -d /tmp/tt/symlink.to.test.folder/
rm: cannot remove '/tmp/tt/symlink.to.test.folder/': Not a directory
`
)

Reply:
So I think the behavior of failing to remove symlink.to.test.folder/ is OK.